kernel/aster-nix/src/fs/ext2/inode.rs

```diff
@@ -161,77 +164,79 @@ impl Inode {
-        drop(inner);
+        let (mut self_inner, mut target_inner) = write_lock_two_inodes(self, &target);
+        // When we got the lock, the dir may have been modified by another thread
+        if self_inner.hard_links() == 0 {
+            return_errno_with_message!(Errno::ENOENT, "dir removed");
+        }
+        let (offset, new_ino) = self_inner
+            .get_entry(name, 0)
+            .map(|(offset, entry)| (offset, entry.ino()))
+            .ok_or(Error::new(Errno::ENOENT))?;
+        if target.ino != new_ino {
+            return_errno!(Errno::ENOENT);
+        }
+        let new_target = self.fs().lookup_inode(target.ino)?;
+        if !Arc::ptr_eq(&target, &new_target) {
+            return_errno!(Errno::ENOENT);
+        }
 
-        inode.inner.write().dec_hard_links();
+        self_inner.remove_entry(name, offset)?;
+        target_inner.dec_hard_links();
         Ok(())
     }
 
     pub fn rmdir(&self, name: &str) -> Result<()> {
-        let self_inner = self.inner.upread();
-        if self_inner.file_type() != FileType::Dir {
-            return_errno!(Errno::ENOTDIR);
-        }
-        if self_inner.hard_links() == 0 {
-            return_errno_with_message!(Errno::ENOENT, "dir removed");
-        }
         if name == "." {
             return_errno_with_message!(Errno::EINVAL, "rmdir on .");
         }
         if name == ".." {
             return_errno_with_message!(Errno::ENOTEMPTY, "rmdir on ..");
         }
-        if name.len() > MAX_FNAME_LEN {
-            return_errno!(Errno::ENAMETOOLONG);
-        }
-
-        let dir_inode = {
-            let ino = self_inner
-                .get_entry_ino(name, 0)
-                .ok_or(Error::new(Errno::ENOENT))?;
-            self.fs().lookup_inode(ino)?
-        };
 
-        // FIXME: There may be a deadlock here.
-        let dir_inner = dir_inode.inner.upread();
+        let dir_inode = self.lookup(name)?;
+        let dir_inner = dir_inode.inner.read();
         if dir_inner.file_type() != FileType::Dir {
             return_errno!(Errno::ENOTDIR);
         }
         if dir_inner.entry_count() > 2 {
             return_errno!(Errno::ENOTEMPTY);
         }
+        drop(dir_inner);
 
-        let mut self_inner = self_inner.upgrade();
-        self_inner.remove_entry(name, 0)?;
-        drop(self_inner);
+        let (mut self_inner, mut dir_inner) = write_lock_two_inodes(self, &dir_inode);
+        // When we got the lock, the dir may have been modified by another thread
+        if self_inner.hard_links() == 0 {
+            return_errno_with_message!(Errno::ENOENT, "dir removed");
+        }
+        let (offset, new_ino) = self_inner
+            .get_entry(name, 0)
+            .map(|(offset, entry)| (offset, entry.ino()))
+            .ok_or(Error::new(Errno::ENOENT))?;
+        if dir_inode.ino != new_ino {
+            return_errno!(Errno::ENOENT);
+        }
+        let new_dir = self.fs().lookup_inode(dir_inode.ino)?;
```

Comment:
This `new_dir` could not be newly created right? If so, `potential_new_dir` is clearer.

Reply:
Thanks for your comments.